[profiler] Use a background thread to send out sampling signals.

alexrp opened the pull request:
Previously, when using an interval timer, the initial profiling signal could be
delivered to *any* thread in the process. This is not normally a problem, but
if the signaled thread has not even finished its initialization inside libc, it
can happen that it hasn't even set up thread-local storage yet. We then blow up
spectacularly when trying to back up errno in the SIGPROF signal handler, as
it's a TLS variable. Even the SIGSEGV handler blows up immediately after as it
can't access JIT TLS data.

Since there appears to be no reliable and portable way for a library like Mono
to control which exact thread gets the initial timer signal, instead switch to
using a background thread that uses a high-resolution sleep and attempts to
switch itself to real time scheduling if possible. This way, we have full
control over which threads we send SIGPROF to, letting us avoid any threads
which aren't in a 'good' state for profiling.

This commit also gets rid of the multiplexing that was going on in the SIGPROF
signal handler, since we now send out the signals from a background thread. As
this was the only use of the async job API, that API has been removed. This
indirectly fixes the signal storming issue that sometimes popped up, where for
some reason multiple threads would think that they're the initiating thread,
resulting in way too many signals going out.

A nice side-effect of doing the signaling ourselves is that we can now use
real-time signals on systems that have them (e.g. Linux), resulting in a nearly
100% signal delivery rate in all cases. Previously, we would lose a tremendous
amount of signals when an application was under heavy load.

alexrp commented:
@kumpera I *think* this covers everything we discussed. Let me know if there's anything that can be improved still.

kumpera commented on mono/mini/mini-posix.c:
```diff
@@ -521,123 +505,255 @@ mono_runtime_cleanup_handlers (void)
-#ifdef HAVE_LINUX_RTC_H
-	if (rtc_fd >= 0)
-		enable_rtc_timer (FALSE);
+	MONO_PREPARE_BLOCKING;
```
Since this is only going to be use by the sampling thread, we should not be getting into the business of coop.

Because it will checkpoint but nobody will even try to suspend it.

alexrp replied:
Good point, those were leftovers from when I was trying to abstract this stuff in `mono-time.c`.

kumpera commented on mono/mini/mini-posix.c:
```diff
@@ -521,123 +505,255 @@ mono_runtime_cleanup_handlers (void)
+	clock_serv_t clock_service;
+	mach_timespec_t then, remain_unused;
+
+	if ((ret = host_get_clock_service (mach_host_self (), mach_clock, &clock_service)) != KERN_SUCCESS)
```
We should cache the clock on startup and not lookup/free on each call.

kumpera commented on mono/mini/mini-posix.c:
```diff
@@ -521,123 +505,255 @@ mono_runtime_cleanup_handlers (void)
+	then.tv_nsec = ns_abs % 1000000000;
+
+	do {
+		ret = clock_sleep (clock_service, TIME_ABSOLUTE, then, &remain_unused);
```
Why are we ignoring the remain time? If we get spuriously woken up, we'd not sleep as requested.

alexrp replied:
Same thing applies here as for `clock_nanosleep`: `clock_sleep` will return immediately if `then` is in the past.

kumpera commented on mono/mini/mini-posix.c:
```diff
@@ -521,123 +505,255 @@ mono_runtime_cleanup_handlers (void)
+	then.tv_nsec = ns_abs % 1000000000;
+
+	do {
+		ret = clock_nanosleep (posix_clock, TIMER_ABSTIME, &then, NULL);
```
You're discarding the last argument. Which must be used to ensure we can sleep the full amount in the case we get spuriously woken up.

alexrp replied:
We don't need to use the `remain` argument. It actually won't even be set since we're passing `TIMER_ABSTIME`. `clock_nanosleep` will simply return immediately if `then` is in the past.

kumpera commented on mono/mini/mini-posix.c:
```diff
@@ -521,123 +505,255 @@ mono_runtime_cleanup_handlers (void)
+	pthread_setschedparam (pthread_self (), SCHED_FIFO, &sched);
+
+	while (InterlockedRead (&sampling_thread_running)) {
+		guint64 sleep = clock_get_time_ns (mode) + rate;
```
This is explicitly introducing signaling drift back.

We should compute the time once and simply add rate on each loop.

alexrp replied:
That's a really good point and does result in much better sleep times.

alexrp commented:
@kumpera I addressed your comments regarding coop macros, clock caching, and sleep time computation.

As mentioned, I'm not quite sure what you mean regarding the last argument of `clock_sleep` / `clock_nanosleep` though.

kumpera commented:
On the remain argument part, this maybe is me being paranoid.

The wording is not 100% clear that sleep won't finish earlier even when using abs time and have remain set to something.

So I think this PR is ready. :+1:

alexrp commented:
build

kumpera commented:
@alexrp this now doesn't merge, can you rebase and merge?

alexrp commented:
OK, rebased, let's see what Jenkins says.

alexrp commented:
@monojenkins merge
